File: verifier-kotlin/src/main/kotlin/tech/thessemaj/deviceintelligence/verifier/Der.kt
package tech.thessemaj.deviceintelligence.verifier

/**
 * A minimal DER TLV reader — just enough to walk the Android Key Attestation
 * extension (KeyDescription) by hand, with no ASN.1 library. Direct port of the
 * helpers in tools/server/verify_token.py (_der_read_len / _read_tlv / _tlv_list).
 *
 * Only structural decoding is done here (tags, lengths, raw values); semantic
 * interpretation lives in [Attestation].
 */
internal object Der {

    /** One decoded TLV: its full tag bytes and its raw value bytes. */
    data class Tlv(val tag: ByteArray, val value: ByteArray)

    /** Read a DER length at [i]; returns (length, nextIndex). */
    private fun readLen(b: ByteArray, i0: Int): Pair<Int, Int> {
        var i = i0
        val n = b[i].toInt() and 0xff; i++
        if (n < 0x80) return n to i
        val k = n and 0x7f
        var v = 0
        var j = 0
        while (j < k) { v = (v shl 8) or (b[i].toInt() and 0xff); i++; j++ }
        return v to i
    }

    /** Read one TLV at [i0] (handles multi-byte high-tag numbers). Returns (tlv, nextIndex). */
    fun readTlv(b: ByteArray, i0: Int): Pair<Tlv, Int> {
        var i = i0
        val start = i
        val t = b[i].toInt() and 0xff; i++
        if ((t and 0x1f) == 0x1f) {                 // high-tag-number form
            while ((b[i].toInt() and 0x80) != 0) i++
            i++
        }
        val tag = b.copyOfRange(start, i)
        val (ln, afterLen) = readLen(b, i)
        i = afterLen
        val value = b.copyOfRange(i, i + ln)
        i += ln
        return Tlv(tag, value) to i
    }

    /** Split a SEQUENCE/SET value into its element TLVs, in order. */
    fun tlvList(seqValue: ByteArray): List<Tlv> {
        val out = ArrayList<Tlv>()
        var i = 0
        while (i < seqValue.size) {
            val (tlv, next) = readTlv(seqValue, i)
            out.add(tlv)
            i = next
        }
        return out
    }

    /** Unwrap one outer SEQUENCE and return its element TLVs. */
    fun sequenceElements(der: ByteArray): List<Tlv> {
        val (outer, _) = readTlv(der, 0)
        return tlvList(outer.value)
    }
}

File: verifiers/node/src/der.ts
// Minimal DER TLV reader (Der.kt port) — high-tag numbers included.
export interface Tlv { tag: Buffer; value: Buffer; }

export function readTlv(b: Buffer, i0: number): [Tlv, number] {
  let i = i0;
  const start = i;
  const t = b[i] & 0xff; i += 1;
  if ((t & 0x1f) === 0x1f) {           // high-tag-number form
    while ((b[i] & 0x80) !== 0) i += 1;
    i += 1;
  }
  const tag = b.subarray(start, i);
  let n = b[i] & 0xff; i += 1;
  let length: number;
  if (n < 0x80) length = n;
  else {
    const k = n & 0x7f;
    let v = 0;
    for (let j = 0; j < k; j++) { v = (v << 8) | (b[i] & 0xff); i += 1; }
    length = v;
  }
  const value = Buffer.from(b.subarray(i, i + length));
  return [{ tag, value }, i + length];
}

export function tlvList(seqValue: Buffer): Tlv[] {
  const out: Tlv[] = [];
  let i = 0;
  while (i < seqValue.length) {
    const [tlv, next] = readTlv(seqValue, i);
    out.push(tlv); i = next;
  }
  return out;
}

export function sequenceElements(der: Buffer): Tlv[] {
  const [outer] = readTlv(der, 0);
  return tlvList(outer.value);
}
